bootloader/src/lib.rs

```diff
@@ -0,0 +1,178 @@
+#![feature(asm)]
+#![feature(link_args)]
+
+use core::panic::PanicInfo;
+use core::*;
+use x86::io::{ inb, outb, insl };
+
+const ELF_MAGIC: u32 = 0x464C457F;
+
+// See https://en.wikipedia.org/wiki/Executable_and_Linkable_Format
+// for detailed breakdown
+#[repr(C)]
+#[derive(Default, Clone)]
+struct elfHeader {
+    magic: u32,       // must contain 0x464C457FU
+    elf: [u8; 12],    // various meta things
+    file_type: u16,   // file object type
+    machine: u16,     // arch, 0x03 is x86
+    version: u32,     // elf version
+    entry: u32,       // entry point for the elf file
+    phoff: u32,       // offset where the program header starts
+    shoff: u32,       // section header offset
+    flags: u32,       // see wikipedia
+    ehsize: u16,      // header length
+    phentsize: u16,   // program header entry size
+    phnum: u16,       // number of program header entries
+    shentsize: u16,   // section header entry size
+    shnum: u16,       // number of section header entries
+    shstrndx: u16,    // index into the section header table that contains the
+                      // section names
+}
+
+// See https://en.wikipedia.org/wiki/Executable_and_Linkable_Format
+// for detailed breakdown
+#[repr(C)]
+#[derive(Default, Clone)]
+struct progHeader {
+  prog_type: u32,     // see wikipedia
+  off: u32,           // offset of the segment in the file image
+  vaddr: u32,         // virtual address of the segment in memory
+  paddr: u32,         // segments physical address
+  filesz: u32,        // size of the segment in file in bytes
+  memsz: u32,         // size of the segment in memory in bytes
+  flags: u32,         // see wikipedia
+  align: u32,         // offset to allign the program on
+}
+
+const SECTSIZE: u32 = 512;
+
+#[panic_handler]
+#[no_mangle]
+pub fn panic(_info: &PanicInfo) -> ! {
+    loop {}
+}
+
+
+// Takes a u32 which represents the location of the elfHeader in memory
+// It then builds a elfHeader on the stack
+// This is unsafe because the caller must ensure "elf_loc" contains an address
+// with a elf_header in it.
+unsafe fn get_elf_header(elf_loc: u32) -> elfHeader {
+    // This should be fairly safe....
+    let elf_ref = mem::transmute::<u32, &mut elfHeader>(elf_loc);
+    elf_ref.clone()
+}
+
+
+// Takes a u32 which represents the location of the progHeader in memory
+// It then builds a elfHeader on the stack
+// This is unsafe because the caller must ensure "prog_loc" contains an address
+// with a elf_header in it.
+unsafe fn get_prog_header(prog_loc: u32) -> progHeader {
+    // This should be fairly safe....
+    let prog_ref = mem::transmute::<u32, &mut progHeader>(prog_loc);
+    prog_ref.clone()
+}
+
+#[no_mangle]
+pub extern "C" fn bootmain() {
+    const ELF_LOC: u32 = 0x10000;
+    readseg(ELF_LOC, 4096, 0);
+    // Copy the elf from the heap to the stack for safe access
+    let elf_header = unsafe{ get_elf_header(ELF_LOC) };
+    if elf_header.magic != ELF_MAGIC {
+        loop {}
```

Comment:
Put a TODO here to find some way to communicate this error. I know that all the uart and vga stuff will be in the kernel proper and that this error condition will ~~probably~~ never happen but for the sake of completeness, I would rather be cognizant of this error path then having to rediscover it if we do something wild down the road.

Reply:
We already discussed this but for the sake of completion on the PR, I tried to do this but writing to the vga more than a couple letters makes the binary too big, it's doable but it would take extra work to load the bootloader that isnt loaded by the BIOS and use extra space.